### ISSUE TITLE
Question: sysrepo and libyang scalability

### ISSUE THREAD

**Issue by igortern** [sysrepo/sysrepo#2843]
Hi,
We have an application that manages PON OLT and ONUs using BBF TR-383, TR-385 YANG models. We use the latest released netopeer2 (2.1.23) and friends. Each ONU requires provisioning of a few TCONTs, GEM ports and VSIs (~10 objects per ONU with many leaves). As we add ONUs the performance grows progressively worse and it can be traced to libyang/sysrepo (generally speaking there is nothing else in our application that depends on the number of ONUs).
To give some idea of the performance degradation progression:
- provisioning ONU # 1 takes <1s
- provisioning ONU # 32 takes ~3s
- provisioning ONU # 64 takes ~12s 

Are there known scalabity problems with sysrepo and are there some parameters that we can fine-tune to improve performance?

Regards,
Igor

**Comment by michalvasko** [comment]
Based on what you said it seems you are working with a lot of data so there may be nothing we can do. You would have to provide some performance data with information about the time/instructions spent in functions so that I could see what exactly you were doing and know for certain whether there is something to improve or not.

**Comment by igortern** [comment]
I'll add some profiling and let you know.
Thanks,
Igor

**Comment by igortern** [comment]
I added timestamp in NP2 logs (in us). Attached is NP2 log when gradually configuring ONUs 10 at a time (sample config is attached). It is clear from the log that whatever SR is doing between 
"SR: Processing "/ietf-netconf:edit-config" "rpc" event"
and publishing "change" event to subscribers takes progressively more time. It takes <1s for the 1st 10 ONUs and >10s for ONUs 51-60. 
Please let me know what additional logs to enable or functions to instrument. As it stands scalability is a show-stopper problem because practically speaking 60 ONUs is a very small configuration. Real-life deployment will have ~2000 ONUs per OLT.
Thanks,
Igor
[ONU_00-09_combinedNE-OLTONU-config.xml.txt](https://github.com/sysrepo/sysrepo/files/8938225/ONU_00-09_combinedNE-OLTONU-config.xml.txt)
[netopeer2.log.txt](https://github.com/sysrepo/sysrepo/files/8938227/netopeer2.log.txt)

**Comment by jktjkt** [comment]
I suggest starting via Linux' `perf`. There's a [nice GUI](https://github.com/KDAB/hotspot) for that, and it comes with excellent tutorials. You don't have to change the source code, and it will show you in which functions the time is spent.

**Comment by igortern** [comment]
Nice tool. Attached is trace info captured when adding ONUs 50-60. bottom-up and top-down views show very different results which I'm still learning to interpret, so attaching both.
Thanks,
Igor
![bottom-up](https://user-images.githubusercontent.com/19907931/174692794-58c55ebf-2a7e-4bca-a325-352244fc6b48.png)
![top-down](https://user-images.githubusercontent.com/19907931/174693000-922cb274-655c-4fcd-b935-72746d3843c1.png)

**Comment by igortern** [comment]
Attached are:
sysrepo + /dev/shm populated with data for the 1st 50 ONUs + xml script that populates the next 10
[ONU_50-59_combinedNE-OLTONU-config.xml.txt](https://github.com/sysrepo/sysrepo/files/9106884/ONU_50-59_combinedNE-OLTONU-config.xml.txt)
[sysrepo.zip](https://github.com/sysrepo/sysrepo/files/9106885/sysrepo.zip)
[dev_shm.zip](https://github.com/sysrepo/sysrepo/files/9106886/dev_shm.zip)

Regards,
Igor

**Comment by flyflick** [comment]
We are also curious about the answer to this issue. What is the latest status?

**Comment by michalvasko** [comment]
Optimizations have lower priority than bugs so there is no status, I have not yet managed to look at all the reported issues.

**Comment by michalvasko** [comment]
After reproducing the validation in `yanglint`, I have found an inefficiency in XPath evaluation that I have fixed, which reduced the total number of executed instructions by 19%.

Other than that, I could not find anything else to improve and validation takes so much time because there are *leafrefs* (28% instructions) and *musts* (55% instructions) that need to be evaluated, summed up in a [file](https://github.com/sysrepo/sysrepo/files/9179301/xpath.txt). So, another optimization may be removing some of those from the YANG files, if at all possible. For instance, you can simply set some leafrefs to `require instance false` when *libyang* will completely skip their validation. The obvious drawback is that it will not be guaranteed that their targets exist.

Next, the performance tools usually do not track waiting for IO, which may actually have the biggest impact. Even though the current datastore implementation is quite efficient for a file, I think it cannot be compared to some other database-based solution. We are planning on implementing such a datastore at some point, which may greatly improve load and store times of changes but that remains to be seen.

Lastly, I assume you are using the `Release` build target of *libyang*, it makes a huge difference for XPath evaluation because we are using lots of safety asserts there.

**Comment by igortern** [comment]
Hi Michal,

Thank you for looking into this. I'll try with code from devel and Release build type. Generally speaking, even if results are twice as good as before, scalability and performance might be the biggest obstacle to netopeer2 practical deployment. As shown in this ticket, we were barely ably to provision 40 ONUs without timeouts and processing time grows faster than linearly.  The practical scale is about 2000 ONUs per OLT. We can live without some advanced features, but not if we can't do the basic provisioning. 

Is there a way to improve 'leafref' and 'must' handling scalability, e.g., use hash tables instead of linked lists, etc.?

Thanks,
Igor

**Comment by igortern** [comment]
I tried the latest release netopeer2 2.1.36 and friends built with -DCMAKE_BUILD_TYPE=Release. There is some 15-20% improvement comparing with the previous run, but it is still far from what we need. edit-config request issued by netopeer2-cli starts timing out after provisioning about 30 ONUs. Request that provisions ONUs 50-59 takes about 8 seconds. We need to be able to provision orders of magnitude more (~2000 ONUs). 
I'll try to removing some "must" clauses and making leafrefs optional, but even if it helps it is not a good long term solution. 
We'll probably need to wait for more efficient datastore implementation that you mentioned.
Thanks,
Igor

**Comment by michalvasko** [comment]
I think I may have a solution for the leafrefs, which will require a new SO version of *libyang* so it may take some time. But those still take only 28% of the instructions executed so if my idea is correct, I would again expect some 20% better performance. 

Optimizing `must` evaluation is a completely different issue, XPath is a complex language and there are already many improvements that make the most common and simple XPaths evaluate in a constant time, for example. But it all depends on the expressions and any more complicated are next to impossible to optimize. So I do not really see a solution with those.

As for the datastore, that is also only theoretical at this point so we will see how that goes once the implementation starts.

**Comment by igortern** [comment]
Thanks Michal.
We'll try moving configuration to schema mount, hopefully it will help.

**Comment by michalvasko** [comment]
Schema mount, why would you do that? That will certainly not improve the performance, on the contrary.

**Comment by igortern** [comment]
Well, scalability is the main reason schema mount is promoted in IETF and BBF. Lets say configuration is n objects per ONU and there are m ONUs. Without schema mount we have a big flat datastore with n x m objects. With schema mount there are m "virtual datastores" with n objects each, I was hoping it will scale better.

Btw, I checked all "must" clauses and leafrefs and it is unclear why performance degrades so much with DS size.
- leafref is a direct reference to another object that appears to be hashed
- all 'must' clauses refer to object in 'near proximity', e.g., 'derived from' interface with a specific type
Do you know why there is such a strong (faster than linear) dependency on the DS size?

1 thing that would be very helpful, perhaps you can add in the next release: a function or compile-time constant that disables all 'must' validation globally. It would be much more maintainable than manually editing a bunch of standard YANG models. Practically speaking, the controller takes care of 'must' validation, so, disabling on the server side is not a huge risk.

Thanks
-Igor

**Comment by michalvasko** [comment]
> With schema mount there are m "virtual datastores" with n objects each, I was hoping it will scale better.

Yes, this should be correct except the last part, not sure why it should scale better, all the data are always validated.

> Btw, I checked all "must" clauses and leafrefs and it is unclear why performance degrades so much with DS size.

If you look at this XPath subexpression as an example `if:interface[if:name=current()]`, the evaluation first puts all the `interface` instances into a set and then evaluates the predicate on each one. It must be computed this way in general but not for relatively simple cases such as this one. So, I have improved the detection of predicates that can be evaluated using hashes and now all of those used in your use-case should be, which reduced the instructions needed by half in my tests. Additionally, you should now get linear complexity of the validation or close to it.

**Comment by igortern** [comment]
Hi Michal,

I tested with devel and ONUs 50-59 now take 4.5s. Ie, faster, but still unacceptably slow. What is interesting is that when the same configuration request is sent again, OK response comes back in a fraction of a second. Doesn't request pass through the same validation regardless of whether there is a change or not? I know that the delay is not in the change_cb because change_cb logs on entrance.

Re schema mount:
>> With schema mount there are m "virtual datastores" with n objects each, I was hoping it will scale better.
>Yes, this should be correct except the last part, not sure why it should scale better, all the data are always validated.
Wouldn't incoming data validated against much smaller "virtual DS"?

Thanks
-Igor

**Comment by michalvasko** [comment]
> I tested with devel and ONUs 50-59 now take 4.5s. Ie, faster, but still unacceptably slow.

That is the price of YANG and its complexity. Now the only relevant optimization could be the new datastore.

> Wouldn't incoming data validated against much smaller "virtual DS"?

Theoretically, any data change can cause any other node in the datastore to become invalid. For schema-mount this could probably be optimized to detect whether the base schema/mounted schema need to be re-validated, but I am not sure how complex that would be to implement. In any case, it likely would not make that much of a difference.

**Comment by igortern** [comment]
Thanks Michal.
Do you know why the same request sent the 2nd time is handled much faster (<1s). Is validation done on the entire request or a delta (which is this case is empty)?

**Comment by michalvasko** [comment]
Like I said, if there are no datastore changes, no validation is performed. If there is any change, full datastore is validated (because it is next to impossible to validate only parts of it).

**Comment by igortern** [comment]
I see. In this case it would be really helpful to have APIs that disable 'must' and 'leafref' validation.
Thanks

**Comment by michalvasko** [comment]
That seems like a rather dangerous functionality to have, one of the core principles of YANG datastores is that their data must be valid. That cannot be guaranteed if they are not validated.

**Comment by igortern** [comment]
Yes, but there is no plan B. In many cases controller also does validation and it alleviates the danger.

**Comment by michalvasko** [comment]
What exactly is this controller, does it also store all the YANG data in its own format? If so, it would be the exact use-case for a custom datastore plugin when the duplicated data would not be stored in *sysrepo* at all.

**Comment by igortern** [comment]
By 'controller' I mean a netconf client that lives in the cloud and manages device running netopeer2

**Comment by michalvasko** [comment]
NETCONF client should generally only send RPCs such as `edit-config` and it does not normally have access to the whole datastore so it could not validate any data.

**Comment by igortern** [comment]
In our case client performs a protocol translation from its NBI to WT-383,385 at SBI. So, the client's view at SBI should be consistent, unless there are bugs. I agree with you that disabling validation is dangerous, but in the moment there is no plan B. It is the only way to make neopeer2 practically usable for netconf/yang managed PON devices. Provisioning for all ONUs follow the same template. We would debug this template with validation on and then switch it off for scale/production.

The good long term solution would be to make validation smarter and only validate a small "affected" subset of the datastore. 
What do you think about the following:
The following discussion is for 2 classes of validators: 'must' and 'leafref'.
- leafref preexisting in DS before edit-config can't break unless the edit-config removes a node
- going through IETF and BBF models, absolute majority of 'must' expression are static, ie, referring to another existing leaves. There is very small number that use functions, like count(). I didn't find any wildcard expressions.

Therefore, I think the following design is possible and can dramatically improve performance:
1) Every node (leaf) has the following lists attached to it, updated when DS is modified: 
- list of nodes that could be affected by the change in "this" node. For example, nodes referring to "this" node via leafref or must/xpath
- list of nodes that could affect "this" node. ie, target nodes "this" node refers to via leafref or must/xpath
2) Some (few) nodes containing non-constant xpath are put onto "always validate" list

When DS is modified, the following validations are done:
- nodes in "always validate" list
For each modified node:
- if the node being updated exists (ie, changed or deleted): go over list of nodes affected by "this" node
- if the node is being added, validate the node itself. not much else is needed

What do I miss here?

**Comment by michalvasko** [comment]
> leafref preexisting in DS before edit-config can't break unless the edit-config removes a node

It is not that simple, a node may be removed indirectly. There could be a `when` expression and even by adding a node you could make it false, which in turn removes this node, which could be the target of a leafref.

> going through IETF and BBF models, absolute majority of 'must' expression are static, ie, referring to another existing leaves. There is very small number that use functions, like count(). I didn't find any wildcard expressions.

That is exactly what I have used to our advantage when I implemented the last optimization of XPath evaluation.

>  Every node (leaf) has the following lists attached to it, updated when DS is modified:
    list of nodes that could be affected by the change in "this" node. For example, nodes referring to "this" node via leafref or must/xpath
    list of nodes that could affect "this" node. ie, target nodes "this" node refers to via leafref or must/xpath

This our first (naive) idea as well, which we implemented in *libyang* v1. All this does in practice is moves the work from validation to data tree changes/freeing. So, the result was that freeing large data trees took insane amounts of time so we decided not to use such a principle in *libyang* v2.

Just imagine what work would be needed when you, let's say, remove a node from a data tree. You would need to go through the whole data tree and remove it from the list of "dependent nodes" of all the other nodes. It can even be optimized, when all the nodes would not only have the list of dependent nodes but also a list of all the nodes that depend on the node. So, in this example you could directly just traverse this other list and remove yourself from all the other nodes dependent lists. However, what about adding a node? You would again need to go through the whole data tree and add the new node into dependent node list of all the relevant nodes.

In short, exactly what I said, while validation would be fast, data modification/freeing would not be. Also, having practical experience with this approach, it adds a lot of complexity into data manipulation, which always need to be approached carefully. Right now, there are no such complex mechanisms and it all works much better.

Finally, note that in your use-case, after my optimizations, must/leafref validation takes only 2/3rds of the whole time required for data validation. Considering there is basically nothing else to validate (but that still needs to be learned), I think that is a good result. It also means that even if we somehow manage to completely get rid of that validation, it would still "only" be a 66% improvement.

**Comment by igortern** [comment]
Thanks Michal

**Comment by michalvasko** [comment]
Just letting you know that the optimizations I thought I had performed were true only for XPath expressions (`must`) but not for leafrefs. So, I made sure they are also evaluated efficiently after which another performance problem was visible, which I managed to fix, too. The result is another huge 73% reduction in the instructions used with the XPath evaluation (`must` and `leafref`) taking only 36%. I do not think I will be able to optimize it significantly in future, this is already a much better result than I expected.

**Comment by igortern** [comment]
Thank you Michal. Provisioning ONUs 50-59 takes 3.5s now
